### PR TITLE
[CAS-1042] Create SuggestionList ViewHolderFactory

### PR DESCRIPTION
**Pull request by bychkovdmitry**
https://stream-io.atlassian.net/browse/CAS-1042

### 🎯 Goal

Create SuggestionList ViewHolder Factory

### 🛠 Implementation details

Added `MessageInputView::setSuggestionListViewHolderFactory` method which allows to provide custom views from suggestion list popup. Click listeners are set internally and not exposed to our customers.

### 🧪 Testing

1. Create `item_command.xml` layout:

```xml
<?xml version="1.0" encoding="utf-8"?>
<FrameLayout xmlns:android="http://schemas.android.com/apk/res/android"
    android:layout_width="match_parent"
    android:layout_height="wrap_content"
    android:background="?attr/selectableItemBackground"
    android:paddingStart="16dp"
    android:paddingEnd="16dp"
    android:paddingTop="8dp"
    android:paddingBottom="8dp">

    <TextView
        android:id="@+id/commandNameTextView"
        android:layout_width="wrap_content"
        android:layout_height="wrap_content" />

</FrameLayout>
```

2. Create custom view holder and view holder factory

```kotlin
class CustomSuggestionListViewHolderFactory : SuggestionListItemViewHolderFactory() {

    override fun createCommandViewHolder(
        parent: ViewGroup,
    ): BaseSuggestionItemViewHolder<SuggestionListItem.CommandItem> {
        return ItemCommandBinding
            .inflate(LayoutInflater.from(parent.context), parent, false)
            .let(::CustomCommandViewHolder)
    }
}

class CustomCommandViewHolder(
    private val binding: ItemCommandBinding,
) : BaseSuggestionItemViewHolder<SuggestionListItem.CommandItem>(binding.root) {

    override fun bindItem(item: SuggestionListItem.CommandItem) {
        binding.commandNameTextView.text = item.command.name
    }
}
```

4. Set custom view holder factory
```kotlin
messageInputView.setSuggestionListViewHolderFactory(CustomSuggestionListViewHolderFactory())
```

### ☑️ Checklist

- [x] I have signed the [Stream CLA](https://docs.google.com/forms/d/e/1FAIpQLScFKsKkAJI7mhCr7K9rEIOpqIDThrWxuvxnwUq2XkHyG154vQ/viewform) (required)
- [x] PR targets the `develop` branch
- [x] Changelog is updated with client-facing changes
- [ ] New code is covered by unit tests
- [x] Comparison screenshots added for visual changes
- [x] Affected documentation updated (CMS, cookbooks, tutorial)
- [x] Reviewers added

### 🎉 GIF

giphy (6)

### CODE REVIEW

**Comment by zsmb13** [comment]
Would be good to add docs to the Docusaurus page of this view about this factory with a simple example

**Comment by bychkovdmitry** [comment]
> Would be good to add docs to the Docusaurus page of this view about this factory with a simple example

Yep, working on it...

**Comment by bychkovdmitry** [comment]
Added Docusaurus docs

File: stream-chat-android-ui-components/src/main/kotlin/io/getstream/chat/android/ui/message/input/MessageInputView.kt [hunk 1]
```diff
@@ -222,12 +228,17 @@ public class MessageInputView : ConstraintLayout {
...
-        styleMentionsIcon(style.mentionsIcon)
+        setSuggestionListViewStyle(
+            SuggestionListViewStyle(
+                suggestionsBackground = style.suggestionsBackground,
+                commandsTitleTextStyle = style.commandsTitleTextStyle,
+                commandsNameTextStyle = style.commandsNameTextStyle,
+                commandsDescriptionStyle = style.commandsDescriptionTextStyle,
+                mentionsUsernameTextStyle = style.mentionsUsernameTextStyle,
+                mentionsNameTextStyle = style.mentionsNameTextStyle,
+                mentionIcon = style.mentionsIcon
```
Comment (adasiewiczr) [inline stream-chat-android-ui-components/src/main/kotlin/io/getstream/chat/android/ui/message/input/MessageInputView.kt:239]:
```suggestion
                mentionIcon = style.mentionsIcon,
```

File: stream-chat-android-ui-components/src/main/kotlin/io/getstream/chat/android/ui/suggestion/list/adapter/SuggestionListItem.kt [hunk 2]
```diff
@@ -0,0 +1,13 @@
+package io.getstream.chat.android.ui.suggestion.list.adapter
+
+import io.getstream.chat.android.client.models.Command
+import io.getstream.chat.android.client.models.User
+import io.getstream.chat.android.ui.suggestion.list.SuggestionListView
+
+/**
+ * [SuggestionListItem] represents elements that are displayed in [SuggestionListView]
```
Comment (adasiewiczr) [inline stream-chat-android-ui-components/src/main/kotlin/io/getstream/chat/android/ui/suggestion/list/adapter/SuggestionListItem.kt:8]:
```suggestion
 * Represents elements that are displayed in [SuggestionListView]
```

File: stream-chat-android-ui-components/src/main/kotlin/io/getstream/chat/android/ui/suggestion/list/adapter/SuggestionListItemViewHolderFactory.kt [hunk 3]
```diff
@@ -0,0 +1,32 @@
+package io.getstream.chat.android.ui.suggestion.list.adapter
+
+import android.view.ViewGroup
+import io.getstream.chat.android.ui.suggestion.list.adapter.viewholder.BaseSuggestionItemViewHolder
+import io.getstream.chat.android.ui.suggestion.list.adapter.viewholder.internal.CommandViewHolder
+import io.getstream.chat.android.ui.suggestion.list.adapter.viewholder.internal.MentionViewHolder
+import io.getstream.chat.android.ui.suggestion.list.internal.SuggestionListViewStyle
+
+public open class SuggestionListItemViewHolderFactory {
```
Comment (adasiewiczr) [inline stream-chat-android-ui-components/src/main/kotlin/io/getstream/chat/android/ui/suggestion/list/adapter/SuggestionListItemViewHolderFactory.kt:9]: Would be really nice to add some KDocs

File: stream-chat-android-ui-components/src/main/kotlin/io/getstream/chat/android/ui/suggestion/list/adapter/SuggestionListItemViewHolderFactory.kt [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+        )
+    }
+
+    public open fun createCommandViewHolder(
+        parentView: ViewGroup,
+    ): BaseSuggestionItemViewHolder<SuggestionListItem.CommandItem> {
+        return CommandViewHolder(
+            parentView,
+            style?.commandsNameTextStyle,
+            style?.commandsDescriptionStyle
```
Comment (adasiewiczr) [inline stream-chat-android-ui-components/src/main/kotlin/io/getstream/chat/android/ui/suggestion/list/adapter/SuggestionListItemViewHolderFactory.kt:29]: Maybe let's use named parameters here and in the method above?

File: stream-chat-android-ui-components/src/main/kotlin/io/getstream/chat/android/ui/suggestion/list/adapter/viewholder/BaseSuggestionItemViewHolder.kt [hunk 5]
```diff
@@ -0,0 +1,15 @@
+package io.getstream.chat.android.ui.suggestion.list.adapter.viewholder
+
+import android.content.Context
+import android.view.View
+import androidx.recyclerview.widget.RecyclerView
+import io.getstream.chat.android.ui.suggestion.list.adapter.SuggestionListItem
+
+public abstract class BaseSuggestionItemViewHolder<T : SuggestionListItem>(itemView: View) :
```
Comment (adasiewiczr) [inline stream-chat-android-ui-components/src/main/kotlin/io/getstream/chat/android/ui/suggestion/list/adapter/viewholder/BaseSuggestionItemViewHolder.kt:8]: Would be nice to add some KDocs

File: stream-chat-android-ui-components/src/main/kotlin/io/getstream/chat/android/ui/suggestion/list/adapter/SuggestionListItemViewHolderFactory.kt [hunk 6]
```diff
@@ -0,0 +1,32 @@
+package io.getstream.chat.android.ui.suggestion.list.adapter
+
+import android.view.ViewGroup
+import io.getstream.chat.android.ui.suggestion.list.adapter.viewholder.BaseSuggestionItemViewHolder
+import io.getstream.chat.android.ui.suggestion.list.adapter.viewholder.internal.CommandViewHolder
+import io.getstream.chat.android.ui.suggestion.list.adapter.viewholder.internal.MentionViewHolder
+import io.getstream.chat.android.ui.suggestion.list.internal.SuggestionListViewStyle
+
+public open class SuggestionListItemViewHolderFactory {
+    internal var style: SuggestionListViewStyle? = null
```
Comment (adasiewiczr) [inline stream-chat-android-ui-components/src/main/kotlin/io/getstream/chat/android/ui/suggestion/list/adapter/SuggestionListItemViewHolderFactory.kt:10]: Is it possible to have a nullable style? Maybe we could convert it to lateinit?

File: stream-chat-android-ui-components/src/main/kotlin/io/getstream/chat/android/ui/suggestion/list/adapter/viewholder/BaseSuggestionItemViewHolder.kt (hunk 5 above)
Comment (adasiewiczr) [reply stream-chat-android-ui-components/src/main/kotlin/io/getstream/chat/android/ui/suggestion/list/adapter/viewholder/BaseSuggestionItemViewHolder.kt:8]: We can do it later

**Review by adasiewiczr — APPROVED** [review]

**Pull request merged**